### ISSUE TITLE
Routing based on content type

### ISSUE THREAD

**Issue by joshtriplett** [http-rs/tide#805]
I'd love to have content-type integrated into routing, allowing something like this:

```rust
app.at("/xyz")
    .content_type("application/x-xyz-request")
    .get(...)
```

This would check that the route matches the specified MIME type, and if not, return a `415 Unsupported Media Type`.

This integration should likely support serving multiple content types for a route, as well.

**Comment by jbr** [comment]
that seems like it could be accomplished with a middleware, like
```rust
app.at("/xyz")
    .with(ContentTypeFilteringMiddleware::only("application/x-xyz-request"))
    .get(...)
```

**Comment by joshtriplett** [comment]
@jbr True, and that would be a substantial improvement. Might also make a good sample middleware.

**Comment by Bauke** [comment]
I've just made [a crate](https://crates.io/crates/tide-content-type-filter) for this, I'm not sure if it's all correct and working in the real world yet but at least mocking it in the tests works.

**Comment by yoshuawuyts** [comment]
@Bauke it could be nice if the filter had full support for `Accept`-like filters so that patterns can be provided into the filter such as `text/*`. [`http_types::Accept::negotiate`](https://docs.rs/http-types/2.12.0/http_types/content/struct.Accept.html#method.negotiate) implements an algorithm for this. Perhaps it could be reused for this purpose too?

**Comment by Bauke** [comment]
I tried out using `Accept` but don't think it could work for filtering, since when the wildcard directive is set it will always return a mime type (so I can't distinguish whether it's allowed or not). I do like the idea of wildcards though so [I added that](https://github.com/Bauke/tide-content-type-filter/commit/3792989b75d7f8ec3d568bf615daf7782e777516) with a simple check if the subtype is a `*` and then only checking for the basetype.
